lib/autoload/course/conditional/conditional_dependency_graph.rb

```diff
@@ -0,0 +1,66 @@
+# Dependency graph to handle and resolve the dependencies for the Conditional-Condition framework.
+class Course::Conditional::ConditionalDependencyGraph
+  private_class_method :new
+
+  def initialize(dependency_graph, edges)
+    @dependency_graph = dependency_graph
+    @edges = edges
+  end
+
+  # Walk through the dependency graph and resolve all conditionals base on the given satisfied
+  # conditions.
+  #
+  # @param[Array<Condition] satisfied_conditions Conditions that have already been satisfied
+  # @param[Course::CourseUser] course_user whose dependencies are to be resolved
+  # @return [Set<Object] All the resolved conditionals base on the given conditions
+  def resolve(satisfied_conditions, course_user)
+    resolved_conditions = Set.new(satisfied_conditions)
+    resolved_conditionals = Set.new
+
+    # Walk through the dependency graph to resolve the conditionals
+    @dependency_graph.each do |conditional|
+      # A conditional is resolved if all its specific conditions are resolved
+      next unless resolved_conditions.superset?(Set.new(conditional.specific_conditions))
+      resolved_conditionals.add(conditional)
+
+      # Add newly resolved conditions to cascade the effect of the resolved conditional
+      resolved_conditions.merge(@edges[conditional].select { |c| c.satisfied_by?(course_user) })
+    end
+
+    resolved_conditionals
+  end
+
+  class << self
+    # Topological sort the given conditionals to produce the conditionals' dependency graph
+    #
+    # @param [Array<Object] conditionals Array of objects with acts_as_conditional
+    # @return [ConditionalDependencyGraph] Topologically sorted conditional dependency graph
+    # @raise [ArgumentError] When there is a cyclic dependency within the given conditionals
+    def build(conditionals)
+      edges = Hash.new { |h, k| h[k] = Set.new }
```

Comment:
Will it be too complex to create a new type in order to wrap an implementation of a hash with default values?